libutils/stack.c

```diff
@@ -0,0 +1,129 @@
+#include <alloc.h>
+#include <logging.h>
+#include <mutex.h>
+#include <stack.h>
+
+#define STACK_BOTTOM 0
+#define EXPAND_FACTOR 2
+
+struct Stack_ {
+    pthread_mutex_t *lock;
+    void **data;
+    int top;
+    int capacity;
+    void (*ItemDestroy) (void *item);
+};
+
+static void DestroyRange(Stack *stack, int start, int end);
+static void ExpandIfNecessary(Stack *stack);
+
+Stack *StackNew(size_t initial_capacity, void (ItemDestroy) (void *item))
+{
+    Stack *stack = xmalloc(sizeof(Stack));
+
+    if (initial_capacity <= 0)
+    {
+        initial_capacity = 1;
```

Comment:
I don't think this is a good default for initial capacity. Small stacks can/should be explicitly requested so I'd set this to at least `16`. That would also make sure that the expansion is not happening too often for the first few items added.

Reply:
+1